src/java.base/share/classes/java/security/PEMEncoder.java

```diff
@@ -0,0 +1,368 @@
+
+/**
+ * PEMEncoder is an immutable class for Privacy-Enhanced Mail (PEM)
+ * data.  PEM is a textual encoding used to store and transfer security
+ * objects, such as asymmetric keys, certificates, and certificate revocation
+ * lists (CRL).  It is defined in RFC 1421 and RFC 7468.  PEM consists of a
+ * Base64-formatted binary encoding enclosed by a type-identifying header
+ * and footer.
+ *
+ * <p> Encoding may be performed on Java Cryptographic Extension (JCE) objects
+ * that implement {@link DEREncodable} and support
+ * {@linkplain PKCS8EncodedKeySpec PKCS#8} or
+ * {@linkplain X509EncodedKeySpec X509} formats.
+ *
+ * <p> Encrypted private key PEM data can be built by calling the encode methods
+ * on a PEMEncoder instance returned by {@link #withEncryption(char[])} or
+ * by passing an {@link EncryptedPrivateKeyInfo} object into the encode methods.
+ *
+ * <p> PKCS8 2.0 allows OneAsymmetricKey encoding, which may contain both private
+ * and public keys in the same PEM.This is supported by using the
+ * {@link KeyPair} class with the encode methods.
+ *
+ * <p> When encoding a {@link PEMRecord}, the API surrounds the
+ * {@linkplain PEMRecord#pem()} with a generated the PEM header and footer
+ * from {@linkplain PEMRecord#type()}.  It will not check the validity of
+ * the data.
+ *
+ * <p>{@code String} values returned by this class use character set
+ * {@link java.nio.charset.StandardCharsets#ISO_8859_1 ISO-8859-1}.
+ *
+ * @apiNote
+ * Here is an example of encoding a PrivateKey object:
+ * {@snippet lang = java:
+ *     PEMEncoder pe = PEMEncoder.of();
+ *     byte[] pemData = pe.encode(privKey);
+ * }
+ *
+ * @see PKCS8EncodedKeySpec
+ * @see X509EncodedKeySpec
+ *
+ * @spec https://www.rfc-editor.org/info/rfc1421
+ *       RFC 1421: Privacy Enhancement for Internet Electronic Mail
+ * @spec https://www.rfc-editor.org/info/rfc7468
+ *       RFC 7468: Textual Encodings of PKIX, PKCS, and CMS Structures
+ *
+ * @since 25
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+public final class PEMEncoder {
+
+    // Singleton instance of PEMEncoder
+    private static final PEMEncoder PEM_ENCODER = new PEMEncoder(null);
+
+    // Stores the password for an encrypted encoder that isn't setup yet.
+    private PBEKeySpec keySpec;
+    // Stores the key after the encoder is ready to encrypt.  The prevents
+    // repeated SecretKeyFactory calls if the encoder is used on multiple keys.
+    private SecretKey key;
+    // Makes SecretKeyFactory generation thread-safe.
+    private final ReentrantLock lock;
+    // Lazy initialize singleton encoder.
+    private static Base64.Encoder b64Encoder;
+
+    /**
+     * Instantiate a new PEMEncoder for Encrypted Private Keys.
+     *
+     * @param pbe contains the password spec used for encryption.
+     */
+    private PEMEncoder(PBEKeySpec pbe) {
+        keySpec = pbe;
+        lock = new ReentrantLock();
+    }
+
+    /**
+     * Returns an instance of PEMEncoder.
```

Comment:
Do we need to say "simple" or "normal" or "raw" `PEMEncoder` here? Just to be clear that it does not do any encryption.